src/core/socket.c

```diff
@@ -2816,6 +2927,19 @@ void socket_connection_unref(Socket *s) {
         assert(s->n_connections > 0);
         s->n_connections--;
 
+        if (service->peer) {
+                service->peer->n_ref--;
```

Comment:
please move this into the usual service_peer_unref() call, as we typically (though not always) do for reference counted objects.